File: task-30bf78/environment/lox/parser.py
import sys
from .tokens import TokenType
from .ast_nodes import (
    Literal, Grouping, Unary, Binary, Variable, Assign, Logical, Call,
    Get, Set, This, Super, ExpressionStmt, PrintStmt, VarStmt, BlockStmt,
    IfStmt, WhileStmt, FunctionStmt, ReturnStmt, ClassStmt,
)
from .errors import LoxParseError


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.current = 0
        self.had_error = False

    def parse(self):
        statements = []
        while not self.is_at_end():
            stmt = self.declaration()
            if stmt is not None:
                statements.append(stmt)
        return statements

    def declaration(self):
        try:
            if self.match(TokenType.CLASS):
                return self.class_declaration()
            if self.match(TokenType.FUN):
                return self.function("function")
            if self.match(TokenType.VAR):
                return self.var_declaration()
            return self.statement()
        except LoxParseError as e:
            print(
                f"[line {e.token.line}] Error at '{e.token.lexeme}': {e.message}",
                file=sys.stderr,
            )
            self.had_error = True
            self.synchronize()
            return None

    def class_declaration(self):
        name = self.consume(TokenType.IDENTIFIER, "Expect class name.")
        superclass = None
        if self.match(TokenType.LESS):
            self.consume(TokenType.IDENTIFIER, "Expect superclass name.")
            superclass = Variable(self.previous())
        self.consume(TokenType.LEFT_BRACE, "Expect '{' before class body.")
        methods = []
        while not self.check(TokenType.RIGHT_BRACE) and not self.is_at_end():
            methods.append(self.function("method"))
        self.consume(TokenType.RIGHT_BRACE, "Expect '}' after class body.")
        return ClassStmt(name, superclass, methods)

    def function(self, kind):
        name = self.consume(TokenType.IDENTIFIER, f"Expect {kind} name.")
        self.consume(TokenType.LEFT_PAREN, f"Expect '(' after {kind} name.")
        params = []
        if not self.check(TokenType.RIGHT_PAREN):
            while True:
                params.append(
                    self.consume(TokenType.IDENTIFIER, "Expect parameter name.")
                )
                if not self.match(TokenType.COMMA):
                    break
        self.consume(TokenType.RIGHT_PAREN, "Expect ')' after parameters.")
        self.consume(TokenType.LEFT_BRACE, f"Expect '{{' before {kind} body.")
        body = self.block()
        return FunctionStmt(name, params, body)

    def var_declaration(self):
        name = self.consume(TokenType.IDENTIFIER, "Expect variable name.")
        initializer = None
        if self.match(TokenType.EQUAL):
            initializer = self.expression()
        self.consume(TokenType.SEMICOLON, "Expect ';' after variable declaration.")
        return VarStmt(name, initializer)

    def statement(self):
        if self.match(TokenType.FOR):
            return self.for_statement()
        if self.match(TokenType.IF):
            return self.if_statement()
        if self.match(TokenType.PRINT):
            return self.print_statement()
        if self.match(TokenType.RETURN):
            return self.return_statement()
        if self.match(TokenType.WHILE):
            return self.while_statement()
        if self.match(TokenType.LEFT_BRACE):
            return BlockStmt(self.block())
        return self.expression_statement()

    def for_statement(self):
        self.consume(TokenType.LEFT_PAREN, "Expect '(' after 'for'.")

        if self.match(TokenType.SEMICOLON):
            initializer = None
        elif self.match(TokenType.VAR):
            initializer = self.var_declaration()
        else:
            initializer = self.expression_statement()

        condition = None
        if not self.check(TokenType.SEMICOLON):
            condition = self.expression()
        self.consume(TokenType.SEMICOLON, "Expect ';' after loop condition.")

        increment = None
        if not self.check(TokenType.RIGHT_PAREN):
            increment = self.expression()
        self.consume(TokenType.RIGHT_PAREN, "Expect ')' after for clauses.")

        body = self.statement()

        if increment is not None:
            body = BlockStmt([body, ExpressionStmt(increment)])

        if condition is None:
            condition = Literal(True)
        body = WhileStmt(condition, body)

        if initializer is not None:
            body = BlockStmt([initializer, body])

        return body

    def if_statement(self):
        self.consume(TokenType.LEFT_PAREN, "Expect '(' after 'if'.")
        condition = self.expression()
        self.consume(TokenType.RIGHT_PAREN, "Expect ')' after if condition.")
        then_branch = self.statement()
        else_branch = None
        if self.match(TokenType.ELSE):
            else_branch = self.statement()
        return IfStmt(condition, then_branch, else_branch)

    def print_statement(self):
        value = self.expression()
        self.consume(TokenType.SEMICOLON, "Expect ';' after value.")
        return PrintStmt(value)

    def expression_statement(self):
        expr = self.expression()
        self.consume(TokenType.SEMICOLON, "Expect ';' after expression.")
        return ExpressionStmt(expr)

    def return_statement(self):
        keyword = self.previous()
        value = None
        if not self.check(TokenType.SEMICOLON):
            value = self.expression()
        self.consume(TokenType.SEMICOLON, "Expect ';' after return value.")
        return ReturnStmt(keyword, value)

    def while_statement(self):
        self.consume(TokenType.LEFT_PAREN, "Expect '(' after 'while'.")
        condition = self.expression()
        self.consume(TokenType.RIGHT_PAREN, "Expect ')' after condition.")
        body = self.statement()
        return WhileStmt(condition, body)

    def block(self):
        statements = []
        while not self.check(TokenType.RIGHT_BRACE) and not self.is_at_end():
            stmt = self.declaration()
            if stmt is not None:
                statements.append(stmt)
        self.consume(TokenType.RIGHT_BRACE, "Expect '}' after block.")
        return statements

    def expression(self):
        return self.assignment()

    def assignment(self):
        expr = self.or_expr()
        if self.match(TokenType.EQUAL):
            equals = self.previous()
            value = self.assignment()
            if isinstance(expr, Variable):
                return Assign(expr.name, value)
            elif isinstance(expr, Get):
                return Set(expr.obj, expr.name, value)
            raise LoxParseError(equals, "Invalid assignment target.")
        return expr

    def or_expr(self):
        expr = self.and_expr()
        while self.match(TokenType.OR):
            operator = self.previous()
            right = self.and_expr()
            expr = Logical(expr, operator, right)
        return expr

    def and_expr(self):
        expr = self.equality()
        while self.match(TokenType.AND):
            operator = self.previous()
            right = self.equality()
            expr = Logical(expr, operator, right)
        return expr

    def equality(self):
        expr = self.comparison()
        while self.match(TokenType.BANG_EQUAL, TokenType.EQUAL_EQUAL):
            operator = self.previous()
            right = self.comparison()
            expr = Binary(expr, operator, right)
        return expr

    def comparison(self):
        expr = self.term()
        while self.match(
            TokenType.GREATER,
            TokenType.GREATER_EQUAL,
            TokenType.LESS,
            TokenType.LESS_EQUAL,
        ):
            operator = self.previous()
            right = self.term()
            expr = Binary(expr, operator, right)
        return expr

    def term(self):
        expr = self.factor()
        while self.match(TokenType.PLUS, TokenType.MINUS):
            operator = self.previous()
            right = self.factor()
            expr = Binary(expr, operator, right)
        return expr

    def factor(self):
        expr = self.unary()
        while self.match(TokenType.STAR, TokenType.SLASH):
            operator = self.previous()
            right = self.unary()
            expr = Binary(expr, operator, right)
        return expr

    def unary(self):
        if self.match(TokenType.BANG, TokenType.MINUS):
            operator = self.previous()
            right = self.unary()
            return Unary(operator, right)
        return self.call()

    def call(self):
        expr = self.primary()
        while True:
            if self.match(TokenType.LEFT_PAREN):
                expr = self.finish_call(expr)
            elif self.match(TokenType.DOT):
                name = self.consume(
                    TokenType.IDENTIFIER, "Expect property name after '.'."
                )
                expr = Get(expr, name)
            else:
                break
        return expr

    def finish_call(self, callee):
        arguments = []
        if not self.check(TokenType.RIGHT_PAREN):
            while True:
                arguments.append(self.assignment())
                if not self.match(TokenType.COMMA):
                    break
        paren = self.consume(TokenType.RIGHT_PAREN, "Expect ')' after arguments.")
        return Call(callee, paren, arguments)

    def primary(self):
        if self.match(TokenType.FALSE):
            return Literal(False)
        if self.match(TokenType.TRUE):
            return Literal(True)
        if self.match(TokenType.NIL):
            return Literal(None)
        if self.match(TokenType.NUMBER, TokenType.STRING):
            return Literal(self.previous().literal)
        if self.match(TokenType.SUPER):
            keyword = self.previous()
            self.consume(TokenType.DOT, "Expect '.' after 'super'.")
            method = self.consume(TokenType.IDENTIFIER, "Expect superclass method name.")
            return Super(keyword, method)
        if self.match(TokenType.THIS):
            return This(self.previous())
        if self.match(TokenType.LEFT_PAREN):
            expr = self.expression()
            self.consume(TokenType.RIGHT_PAREN, "Expect ')' after expression.")
            return Grouping(expr)
        if self.match(TokenType.IDENTIFIER):
            return Variable(self.previous())
        raise LoxParseError(self.peek(), "Expect expression.")

    def match(self, *types):
        for ttype in types:
            if self.check(ttype):
                self.advance()
                return True
        return False

    def check(self, ttype):
        if self.is_at_end():
            return False
        return self.peek().ttype == ttype

    def advance(self):
        if not self.is_at_end():
            self.current += 1
        return self.previous()

    def is_at_end(self):
        return self.peek().ttype == TokenType.EOF

    def peek(self):
        return self.tokens[self.current]

    def previous(self):
        return self.tokens[self.current - 1]

    def consume(self, ttype, message):
        if self.check(ttype):
            return self.advance()
        raise LoxParseError(self.peek(), message)

    def synchronize(self):
        self.advance()
        while not self.is_at_end():
            if self.previous().ttype == TokenType.SEMICOLON:
                return
            if self.peek().ttype in (
                TokenType.CLASS,
                TokenType.FUN,
                TokenType.VAR,
                TokenType.FOR,
                TokenType.IF,
                TokenType.WHILE,
                TokenType.PRINT,
                TokenType.RETURN,
            ):
                return
            self.advance()

File: bids-expression-evaluator-292ea5/environment/src/parser.ts
import { Token, TokenType, Expr } from './types';

export function parse(tokens: Token[]): Expr {
  let pos = 0;

  function peek(): Token { return tokens[pos]; }
  function advance(): Token { return tokens[pos++]; }

  function expect(type: TokenType): Token {
    const t = peek();
    if (t.type !== type) {
      throw new Error(`Expected ${type} but got ${t.type} ('${t.value}') at position ${t.pos}`);
    }
    return advance();
  }

  function match(...types: TokenType[]): Token | null {
    if (types.includes(peek().type)) return advance();
    return null;
  }

  function parseExpr(): Expr {
    return parseOr();
  }

  function parseOr(): Expr {
    let left = parseAnd();
    while (match(TokenType.Or)) {
      left = { kind: 'binary', op: '||', left, right: parseAnd() };
    }
    return left;
  }

  function parseAnd(): Expr {
    let left = parseEquality();
    while (match(TokenType.And)) {
      left = { kind: 'binary', op: '&&', left, right: parseEquality() };
    }
    return left;
  }

  function parseEquality(): Expr {
    let left = parseComparison();
    let t: Token | null;
    while ((t = match(TokenType.EqualEqual, TokenType.BangEqual))) {
      left = { kind: 'binary', op: t.value, left, right: parseComparison() };
    }
    return left;
  }

  function parseComparison(): Expr {
    let left = parseIn();
    let t: Token | null;
    while ((t = match(TokenType.Less, TokenType.LessEqual, TokenType.Greater, TokenType.GreaterEqual))) {
      left = { kind: 'binary', op: t.value, left, right: parseIn() };
    }
    return left;
  }

  function parseIn(): Expr {
    let left = parseAddition();
    if (match(TokenType.In)) {
      left = { kind: 'binary', op: 'in', left, right: parseAddition() };
    }
    return left;
  }

  function parseAddition(): Expr {
    let left = parseMultiplication();
    let t: Token | null;
    while ((t = match(TokenType.Plus, TokenType.Minus))) {
      left = { kind: 'binary', op: t.value, left, right: parseMultiplication() };
    }
    return left;
  }

  function parseMultiplication(): Expr {
    let left = parseUnary();
    let t: Token | null;
    while ((t = match(TokenType.Star, TokenType.Slash, TokenType.Percent))) {
      left = { kind: 'binary', op: t.value, left, right: parseUnary() };
    }
    return left;
  }

  function parseUnary(): Expr {
    if (match(TokenType.Bang)) {
      return { kind: 'unary', op: '!', operand: parseUnary() };
    }
    if (match(TokenType.Minus)) {
      return { kind: 'unary', op: '-', operand: parseUnary() };
    }
    return parsePostfix();
  }

  function parsePostfix(): Expr {
    let expr = parsePrimary();
    while (true) {
      if (match(TokenType.Dot)) {
        const prop = expect(TokenType.Identifier);
        expr = { kind: 'member', object: expr, property: prop.value };
      } else if (peek().type === TokenType.LeftBracket) {
        advance();
        const idx = parseExpr();
        expect(TokenType.RightBracket);
        expr = { kind: 'index', object: expr, index: idx };
      } else {
        break;
      }
    }
    return expr;
  }

  function parsePrimary(): Expr {
    const t = peek();

    if (t.type === TokenType.Number) {
      advance();
      return { kind: 'literal', value: parseFloat(t.value) };
    }

    if (t.type === TokenType.String) {
      advance();
      return { kind: 'literal', value: t.value };
    }

    if (t.type === TokenType.True) {
      advance();
      return { kind: 'literal', value: true };
    }

    if (t.type === TokenType.False) {
      advance();
      return { kind: 'literal', value: false };
    }

    if (t.type === TokenType.Null) {
      advance();
      return { kind: 'literal', value: null };
    }

    if (t.type === TokenType.Identifier) {
      advance();
      // Check for function call
      if (peek().type === TokenType.LeftParen) {
        advance(); // consume '('
        const args: Expr[] = [];
        if (peek().type !== TokenType.RightParen) {
          args.push(parseExpr());
          while (match(TokenType.Comma)) {
            args.push(parseExpr());
          }
        }
        expect(TokenType.RightParen);
        return { kind: 'call', name: t.value, args };
      }
      return { kind: 'identifier', name: t.value };
    }

    if (t.type === TokenType.LeftParen) {
      advance();
      const expr = parseExpr();
      expect(TokenType.RightParen);
      return expr;
    }

    if (t.type === TokenType.LeftBracket) {
      advance();
      const elements: Expr[] = [];
      if (peek().type !== TokenType.RightBracket) {
        elements.push(parseExpr());
        while (match(TokenType.Comma)) {
          elements.push(parseExpr());
        }
      }
      expect(TokenType.RightBracket);
      return { kind: 'array', elements };
    }

    if (t.type === TokenType.LeftBrace) {
      advance();
      expect(TokenType.RightBrace);
      return { kind: 'object' };
    }

    throw new Error(`Unexpected token ${t.type} ('${t.value}') at position ${t.pos}`);
  }

  const result = parseExpr();
  if (peek().type !== TokenType.EOF) {
    const t = peek();
    throw new Error(`Unexpected token after expression: ${t.type} ('${t.value}') at position ${t.pos}`);
  }
  return result;
}
